File: tonghua-project/backend/app/schemas/ai_assistant.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class AIChatMessage(BaseModel):
    role: str = Field(..., pattern="^(user|assistant|system)$")
    content: str


class AIChatRequest(BaseModel):
    session_id: str = Field(..., max_length=100)
    message: str = Field(..., min_length=1, max_length=2000)
    interaction_type: str = Field(
        "chat",
        pattern="^(chat|product_recommendation|sustainability_advice|donation_guidance|after_sales_help|style_matching)$"
    )
    context: Optional[Dict[str, Any]] = None    # 上下文数据（产品ID、订单ID等）
    history: Optional[List[AIChatMessage]] = Field(None, max_length=20)


class AIChatResponse(BaseModel):
    session_id: str
    message: str
    interaction_type: str
    suggestions: Optional[List[Dict[str, Any]]] = None   # 推荐产品/活动等
    actions: Optional[List[Dict[str, str]]] = None       # 可执行操作


class AIFeedbackRequest(BaseModel):
    interaction_id: int
    feedback: str = Field(..., pattern="^(helpful|not_helpful)$")


class ClothingDonationCreate(BaseModel):
    clothing_type: str = Field(..., max_length=100)
    quantity: int = Field(1, ge=1, le=100)
    condition: str = Field(..., pattern="^(new|like_new|good|fair)$")
    description: Optional[str] = None
    images: Optional[List[str]] = None
    pickup_address: Optional[str] = None
    pickup_time_slot: Optional[str] = Field(None, max_length=100)
    campaign_id: Optional[int] = None


class ClothingDonationUpdate(BaseModel):
    status: Optional[str] = Field(
        None,
        pattern="^(submitted|scheduled|picked_up|processing|converted|completed|rejected)$"
    )
    converted_product_id: Optional[int] = None
    admin_note: Optional[str] = None
    pickup_time_slot: Optional[str] = Field(None, max_length=100)


class ClothingDonationOut(BaseModel):
    id: int
    user_id: int
    campaign_id: Optional[int] = None
    clothing_type: str
    quantity: int
    condition: str
    description: Optional[str] = None
    images: Optional[List[str]] = None
    pickup_address: Optional[str] = None
    pickup_time_slot: Optional[str] = None
    status: str
    converted_product_id: Optional[int] = None
    admin_note: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class SustainabilityMetricOut(BaseModel):
    id: int
    entity_type: str
    entity_id: int
    carbon_saved_kg: Optional[float] = None
    water_saved_liters: Optional[float] = None
    textile_recycled_kg: Optional[float] = None
    trees_equivalent: Optional[float] = None
    sustainability_score: Optional[float] = None
    certification: Optional[str] = None
    notes: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class SustainabilityMetricCreate(BaseModel):
    entity_type: str = Field(..., pattern="^(product|donation|order)$")
    entity_id: int
    carbon_saved_kg: Optional[float] = Field(None, ge=0)
    water_saved_liters: Optional[float] = Field(None, ge=0)
    textile_recycled_kg: Optional[float] = Field(None, ge=0)
    trees_equivalent: Optional[float] = Field(None, ge=0)
    sustainability_score: Optional[float] = Field(None, ge=0, le=100)
    certification: Optional[str] = Field(None, max_length=200)
    notes: Optional[str] = None

File: tonghua-project/frontend/web-react/src/services/aiAssistant.ts
import api from './api';

export interface AIChatMessage {
  role: 'user' | 'assistant' | 'system';
  content: string;
}

export interface AIChatRequest {
  session_id: string;
  message: string;
  interaction_type?: string;
  context?: Record<string, unknown>;
  history?: AIChatMessage[];
}

export interface AISuggestion {
  type: string;
  label: string;
  path?: string;
}

export interface AIChatResponse {
  session_id: string;
  interaction_id: number | null;
  message: string;
  interaction_type: string;
  suggestions: AISuggestion[] | null;
  actions: Array<{ label: string; path: string }> | null;
}

export interface ClothingDonation {
  id: number;
  user_id: number;
  campaign_id: number | null;
  clothing_type: string;
  quantity: number;
  condition: string;
  description: string | null;
  images: string[];
  pickup_address: string | null;
  pickup_time_slot: string | null;
  status: string;
  converted_product_id: number | null;
  admin_note: string | null;
  created_at: string;
  updated_at: string;
}

export interface SustainabilityMetric {
  entity_type: string;
  entity_id: number;
  carbon_saved_kg: number | null;
  water_saved_liters: number | null;
  textile_recycled_kg: number | null;
  trees_equivalent: number | null;
  sustainability_score: number | null;
  certification: string | null;
  is_estimated?: boolean;
}

export interface SustainabilitySummary {
  total_carbon_saved_kg: number;
  total_water_saved_liters: number;
  total_textile_recycled_kg: number;
  total_items: number;
  children_helped?: number;
  artworks_created?: number;
}

export const sendChatMessage = async (data: AIChatRequest): Promise<AIChatResponse> => {
  const response = await api.post('/ai/chat', data);
  return response.data.data;
};

export const submitAIFeedback = async (interactionId: number, feedback: 'helpful' | 'not_helpful') => {
  const response = await api.post('/ai/feedback', { interaction_id: interactionId, feedback });
  return response.data.data;
};

export const getAIRecommendations = async (context = 'general', limit = 4) => {
  const response = await api.get('/ai/recommend', { params: { context, limit } });
  return response.data.data;
};

export const submitClothingDonation = async (data: {
  clothing_type: string;
  quantity: number;
  condition: string;
  description?: string;
  images?: string[];
  pickup_address?: string;
  pickup_time_slot?: string;
  campaign_id?: number;
}): Promise<ClothingDonation> => {
  const response = await api.post('/ai/clothing-donations', data);
  return response.data.data;
};

export const getMyClothingDonations = async (page = 1, pageSize = 20) => {
  const response = await api.get('/ai/clothing-donations/mine', {
    params: { page, page_size: pageSize },
  });
  return response.data;
};

export const getSustainabilityMetric = async (
  entityType: string,
  entityId: number
): Promise<SustainabilityMetric> => {
  const response = await api.get(`/ai/sustainability/${entityType}/${entityId}`);
  return response.data.data;
};

export const getSustainabilitySummary = async (): Promise<SustainabilitySummary> => {
  const response = await api.get('/ai/sustainability/summary');
  return response.data.data;
};

export const clothingConditionMap: Record<string, string> = {
  new: '全新',
  like_new: '近全新',
  good: '良好',
  fair: '一般',
};

export const clothingDonationStatusMap: Record<string, string> = {
  submitted: '已提交',
  scheduled: '已安排取件',
  picked_up: '已取件',
  processing: '处理中',
  converted: '已转化为商品',
  completed: '完成',
  rejected: '不符合标准',
};
